packages/http-server/src/http-server.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/http-server
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {createServer, Server, ServerRequest, ServerResponse} from 'http';
+import {AddressInfo} from 'net';
+import * as pEvent from 'p-event';
+
+export type HttpRequestListener = (
+  req: ServerRequest,
+  res: ServerResponse,
+) => void;
+
+/**
+ * Object for specifyig the HTTP / HTTPS server options
+ */
+export type HttpServerOptions = {
+  port?: number;
+  host?: string;
+  protocol?: HttpProtocol;
+};
+
+export type HttpProtocol = 'http' | 'https'; // Will be extended to `http2` in the future
+
+/**
+ * HTTP / HTTPS server used by LoopBack's RestServer
+ *
+ * @export
+ * @class HttpServer
+ */
+export class HttpServer {
+  private _port: number;
+  private _host?: string;
+  /**
+   * Protocol, default to `http`
+   */
+  private _protocol: HttpProtocol;
+  private _address: AddressInfo;
+  private httpRequestListener: HttpRequestListener;
+  private httpServer: Server;
+
+  /**
+   * @param httpServerOptions
+   * @param httpRequestListener
+   */
+  constructor(
+    httpRequestListener: HttpRequestListener,
+    httpServerOptions?: HttpServerOptions,
+  ) {
+    this.httpRequestListener = httpRequestListener;
+    if (!httpServerOptions) httpServerOptions = {};
+    this._port = httpServerOptions.port || 0;
+    this._host = httpServerOptions.host || undefined;
+    this._protocol = httpServerOptions.protocol || 'http';
+  }
+
+  /**
+   * Starts the HTTP / HTTPS server
+   */
+  public async start() {
+    this.httpServer = createServer(this.httpRequestListener);
+    this.httpServer.listen(this._port, this._host);
+    await pEvent(this.httpServer, 'listening');
+    this._address = this.httpServer.address() as AddressInfo;
+    this._host = this._host || this._address.address;
+    this._port = this._address.port;
+  }
+
+  /**
+   * Stops the HTTP / HTTPS server
+   */
+  public async stop() {
+    this.httpServer.close();
+    await pEvent(this.httpServer, 'close');
+  }
+
+  /**
+   * Protocol of the HTTP / HTTPS server
+   */
+  public get protocol(): HttpProtocol {
```

Comment:
Why bother to have `getter` for these properties? Do you want to enforce `readonly`?

Reply:
Have you tried `readonly` keyword from TypeScript?

Reply:
Thanks, will use that.

Reply:
I think we can use `readonly` modifier for all properties excluding `port`, because `port` changes as a result of `start()` call. OTOH, maybe we should keep a readonly `_port` property to hold the requested port (e.g. `0`) and then write a `port` getter that will either return the actual port reported by `this._httpServer.address().port` or fall back to `this._port`.

This has an important benefit that multiple start/stop cycles of a server configured to listen on port 0 will always pick a new port. At the moment, I believe subsequent `start()` calls will pick the temporary port number assigned by the operating system for the first `start()` call.

@hacksparrow Please add a unit-test to verify the behavior of our server in this edge case.

Reply:
@raymondfeng even an undefined `host` will behave like `port` set to `0`. If the `host` value is not set during instantiation, the `HttpServer` instance will store it as `undefined`, however, the underlying `http`/`https` will generate a host value for the server - something like `::`.

Since `url` is dependent on `host` and `port`, the only property that can be implemented as a `readonly` would be `protocol`.

@bajtos I will add tests for both `host` and `port`.

Reply:
Added.